api/v1beta2/types.go

```diff
@@ -134,12 +134,42 @@ var (
 	ResourceTypeVPC = ResourceType("vpc")
 	// ResourceTypeSubnet is VPC subnet resource.
 	ResourceTypeSubnet = ResourceType("subnet")
+	// ResourceTypePublicGateway is a VPC Public Gateway resource.
+	ResourceTypePublicGateway = ResourceType("publicGateway")
+	// ResourceTypeSecurityGroup is a VPC Security Group resource.
+	ResourceTypeSecurityGroup = ResourceType("securityGroup")
 	// ResourceTypeCOSInstance is IBM COS instance resource.
 	ResourceTypeCOSInstance = ResourceType("cosInstance")
 	// ResourceTypeResourceGroup is IBM Resource Group.
 	ResourceTypeResourceGroup = ResourceType("resourceGroup")
 )
 
+// COSInstanceReference represents an IBM Cloud COS instance.
+type COSInstanceReference struct {
+	// name defines the name of the COS Instance.
+	// +kubebuilder:validation:MinLength:=3
+	// +kubebuilder:validation:MaxLength:=63
+	// +kubebuilder:validation:Pattern=`^[a-z0-9][a-z0-9.-]{1,61}[a-z0-9]$`
+	// +required
+	Name string `json:"name"`
+
+	// id is the ID of the COS instance.
+	// +optional
+	ID *string `json:"id,omitempty"`
```

Comment:
do we really need both Name and ID field?

Reply:
We usually have both Name and ID fields for resources, just incase want to speed up the look up when ID is known.

Reply:
I can remove, I simply copied the COSInstance from PowerVS code, as is.
If you have found you don't need name, I will remove it here as well.

Reply:
Looks like the current PowerVS has no ID's, I kind of favor Id's over names and currently expect CAPI will not be creating COS Instances or resources....so I will see about using Id's over names ATM.
Will update.

Reply:
I dropped `Name` for now.

Reply:
@cjschaef It all depends on how you are going to setup the relevant controller code, will it need ID or just name or both or you can deduce ID from name in code itself or vice versa. We will leave taking call on which fields to have and not upto you however having clarity on how we plan to use them in future controller code helps us in reviewing/understanding better the entire flow.
/cc @Karthik-K-N 

Note - if we need both cant we reuse vpcResource struct for say cos and bucket obj? (just a thought) @Karthik-K-N your thoughts on this?

Reply:
We cannot drop Name as thats the only way user can specify the Name of COSInstance.

Also It wont be good to add VPCResource to COSInstance as both are two different entities in IBM Cloud. May be if we would rename VPCResource to something else to more appropriate as its just represents ID and Name in general, then only we can do.

Reply:
My expectation at this time, is that the controller will not be creating COS resources, but it may be expected to create a VPC Custom Image, using a qcow2 file in a COS bucket, hence the need for this COS struct. I could rely only on a COS bucket CRN, but if this resource can be re-used, I expected Name/ID would be more beneficial, fields set depending on the requirements (PowerVS vs. VPC, etc.)

However, this is my best assumption on the workflow, as limitations may require changes and I am unsure if PowerVS creates COS resources (bucket maybe?) for the VSI Image. So now I wonder if VPC should deviate from the flow PowerVS uses.
It is early on in development, so I can mirror the flow PowerVS uses, for consistency, which would mean I follow similar field requirements (Name's)?

Perhaps you have an opinion on whether VPC should follow similar processes/expectations as PowerVS, when it makes sense. I am not opposed to relying on the controller to create COS resources if necessary, but I had expected the controller to rely on an existing COS Instance and Bucket and ideally the qcow2 Object holding the RHCOS image, and not be responsible for creating it.

Reply:
I think this is what causing confusion, There are two things
1. In PowerVS we create cos instance and bucket to [upload the ignition](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/cloud/scope/powervs_machine.go#L398) data as a object and pass the object URL as [userdata](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/cloud/scope/powervs_machine.go#L282) while creating machine.

2. We have ibmpowervs_image_controller which takes care of importing the RHCOS image from respective COS bucket to where redhat will push the image into the PowerVS service instance.

Reply:
The question I guess I still have is, do the COS instance and bucket need to be created as a prerequisite? I hadn't seen anything in the CAPI code that creates the instance and the bucket (only the Object/File), but I certainly could have missed that.

If this is the case, and I would follow the same flow as PowerVS (create the COS instance and bucket outside CAPI), that will influence what values I want, although I have seen I am limited by the IBM COS SDK, as far as values as well.

Reply:
We do create COS instance and bucket as a part of IBMPowerVSCluster reconciler https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/2a0c2811d02cd003c56327e485c54e262057c408/controllers/ibmpowervscluster_controller.go#L200-L203